lang/funcs/collection.go

```diff
@@ -106,6 +107,249 @@ var LengthFunc = function.New(&function.Spec{
+		}
+
+		if len(outputList) == 0 {
+			return cty.ListValEmpty(cty.String), nil
+		}
+
+		return cty.ListVal(outputList), nil
+	},
+})
+
+// ContainsFunc contructs a function that determines whether a given list contains
+// a given single value as one of its elements.
+var ContainsFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.List(cty.DynamicPseudoType),
+		},
+		{
+			Name: "value",
+			Type: cty.DynamicPseudoType,
+		},
+	},
+	Type: function.StaticReturnType(cty.Bool),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+
+		_, err = Index(args[0], args[1])
+		if err != nil {
+			return cty.False, nil
+		}
+
+		return cty.True, nil
+	},
+})
+
+// IndexFunc contructs a function that finds the element index for a given value in a list.
+var IndexFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.DynamicPseudoType,
+		},
+		{
+			Name: "value",
+			Type: cty.DynamicPseudoType,
+		},
+	},
+	Type: function.StaticReturnType(cty.Number),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		if !(args[0].Type().IsListType() || args[0].Type().IsTupleType()) {
+			return cty.NilVal, fmt.Errorf("argument must be a list or tuple")
+		}
+
+		if args[0].LengthInt() == 0 { // Easy path
+			return cty.NilVal, fmt.Errorf("Cannot search an empty list")
+		}
+
+		for it := args[0].ElementIterator(); it.Next(); {
+			i, v := it.Element()
+			eq, err := stdlib.Equal(v, args[1])
+			if err != nil {
+				return cty.NilVal, err
+			}
+			if !eq.IsKnown() {
+				return cty.UnknownVal(cty.Number), nil
+			}
+			if eq.True() {
+				return i, nil
+			}
+		}
+		return cty.NilVal, fmt.Errorf("item not found")
+
+	},
+})
+
+// DistinctFunc contructs a function that takes a list and returns a new list
+// with any duplicate elements removed.
+var DistinctFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.List(cty.DynamicPseudoType),
+		},
+	},
+	Type: function.StaticReturnType(cty.List(cty.DynamicPseudoType)),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+		if len(args) != 1 {
+			return cty.NilVal, fmt.Errorf("distinct accepts only one argument")
+		}
+
+		var list []cty.Value
+
+		for it := args[0].ElementIterator(); it.Next(); {
+			_, v := it.Element()
+			list, err = appendIfMissing(list, v)
+			if err != nil {
+				return cty.NilVal, err
+			}
+		}
+
+		return cty.ListVal(list), nil
+	},
+})
+
+// ChunklistFunc contructs a function that splits a single list into fixed-size chunks,
+// returning a list of lists.
+var ChunklistFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.List(cty.DynamicPseudoType),
+		},
+		{
+			Name: "size",
+			Type: cty.Number,
+		},
+	},
+	Type: function.StaticReturnType(cty.List(cty.DynamicPseudoType)),
```

Comment:
I think in this situation it'd be best to have a custom implementation here, so we can return a more precise type and thus give the type checker more information to work with. Something like this, maybe?

```go
Type: func(args []cty.Value) (cty.Type, error) {
    return cty.List(args[0].Type())
}
```

This'd mean that if the caller gives a list of strings, the return type will be a list of lists of strings, and then we can still type-check the result even if the list value is not actually known yet. (e.g. with something like `chunklist(aws_instance.foo.*.id, 2)` where the instances haven't been created yet.)

`cty.List(cty.DynamicPseudoType)` is valid as a return type for this, but it would mean that the result could not be type-checked until apply time.